ystia/yorc
# Feature/gh 131 bootstrap yorc

# Pull Request description

## Description of the change

Added a `yorc bootstrap` command to bootstrap the full stack from Alien4Cloud to Yorc and dependencies on different types of infrastructures: Google Cloud, AWS, OpenStack, Hosts Pool.

### What I did

Under `commands/bootstrap` directory, added the following code:
  * `bootstrap.go` is the command line entry point
  * `inputs.go` deals with inputs, provided interactively, through CLI options, environment variables, or input file
  * `resources.go` deals with resources (tosca components needed to deploy the full stack) bundled in Yorc binary, or passed in argument through the advandced option --resources_zip 
  * `setup.go` installs a local Yorc and Consul to bootstrap the deployment on remote infrastructure
  * `topology.go` generates a topology template from user inputs: which infratructure, single node or HA deployment, etc...
  * `deployment.go` deploys the template and follows steps progress by default, or logs


### How to verify it

Verify that your host is fulfilling prerequisites described at https://github.com/ystia/yorc/blob/feature/GH-131-bootstrap-yorc/doc/bootstrap.rst (either ansible is installed, or you should run within a python virtualenv).

Download/unzip yorc binary built from this Pull Request.

Bootstrap a setup on OpenStack in interactive mode. This can be done running:

```
$ ./yorc bootstrap
```

But as there are usually issues with proxies in our lab, just use a config file `values.yaml` to download dependencies from our pulp repository, creating a yaml file like this one by just replacing <repo ip> but the pulp server IP address :

```yaml
alien4cloud:
  download_url: http://<repo ip>/starlings/alien4cloud/alien4cloud-premium-dist-2.1.0-SM7-dist.tar.gz
  port: 8088
consul:
  download_url: http://<repo ip>/ystia/yorc-deploy/consul_1.2.3_linux_amd64.zip
  port: 8500
terraform:
  download_url: http://<repo ip>/ystia/yorc-deploy/terraform_0.11.8_linux_amd64.zip
  plugins_download_urls:
  - http://<repo ip>/ystia/yorc-deploy/terraform-provider-null_1.0.0_linux_amd64.zip
  - http://<repo ip>/ystia/yorc-deploy/terraform-provider-consul_2.1.0_linux_amd64.zip
  - http://<repo ip>/ystia/yorc-deploy/terraform-provider-openstack_1.9.0_linux_amd64.zip
  - http://<repo ip>/ystia/yorc-deploy/terraform-provider-aws_1.36.0_linux_amd64.zip
  - http://<repo ip>/ystia/yorc-deploy/terraform-provider-google_1.18.0_linux_amd64.zip
jdk:
  download_url: http://<repo ip>/ystia/java-a4c/jdk-8u131-linux-x64.tar.gz
  version: 1.8.0-131-b11
``` 
 
and run this command:

```
$ ./yorc bootstrap --values values.yaml --review
```

Select `OpenStack`, then answer to questions.
You will need to provide details about:
  * the OpenStack infrastructure
  * compute nodes to create on demand
  * the user to use to connect to these compute nodes created on demand
  * the floating ip address to create on demand.
  
Finally you will be asked to review and possibly update all settings (because you specified `--review` above),
and the deployment will start.

When the deployment has finished, login on the Alien4Cloud URL, go to the administration menu, check a Yorc orchestrator was created, with an OpenStack location, and on-demand resources as you specified.

Undeploy this deployment running :

```
$ ./yorc d undeploy -p <deployment ID>
```

Follow the undeployment, running :

```
$ ./yorc d info -f <deployment ID>
```

Once the undeployment is done, cleanup the local environment running :

```
./yorc bootstrap cleanup
```

On a host which has ssh access to Google Cloud Compute Engine (so not an OpenStack VM host in our lab), bootstrap a full stack running
(change values, which are dummy values here) :

```
$ export YORC_INFRA_GOOGLE_APPLICATION_CREDENTIALS=/path/to/serviceaccountcreds.json
$ export YORC_INFRA_GOOGLE_PROJECT=myproject
$ export YORC_ADDRESS_REGION=europe-west1
$ export YORC_COMPUTE_IMAGE_PROJECT=centos-cloud
$ export YORC_COMPUTE_IMAGE_FAMILY=centos-7
$ export YORC_COMPUTE_MACHINE_TYPE=n1-standard-2
$ export YORC_COMPUTE_ZONE=europe-west1-b
$ export YORC_COMPUTE_METADATA="ssh-keys=yorcuser:public key content"
$ export YORC_COMPUTE_TAGS=a4c
$ export YORC_CREDENTIALS_USER=yorcuser
$ export YORC_YORC_PRIVATE_KEY_FILE=/path/to/yorc.pem

$ ./yorc bootstrap
```

As described above, wait for the end of deployment, check the Alien4Cloud URL and config of orchestrator, location, on-demand resources

Undeploy, cleanup the bootstrap environment.

On a host which has ssh access to AWS EC2 (so not an OpenStack VM host in our lab), bootstrap a full stack running
(change values, which are dummy values here) :

```
$ ./yorc bootstrap \
  --infrastructure_aws_region us-east-2 \
  --infrastructure_aws_access_key ABCDEFG \
  --infrastructure_aws_secret_key abcd/efgh \
  --compute_image_id ami-18f8df7d \
  --compute_instance_type t2.large \
  --compute_key_name yourKname \
  --compute_security_groups janus-securityGroup \
  --yorc_private_key_file /path/to/yorc.pem
```

As described above, wait for the end of deployment, check the Alien4Cloud URL and config of orchestrator, location, on-demand resources

Undeploy, cleanup the bootstrap environment.

## Applicable Issues

https://github.com/ystia/yorc/issues/131
